File: app/src/main/java/org/neidhardt/dynamicsoundboard/soundsheetmanagement/model/SoundSheetsManager.java
package org.neidhardt.dynamicsoundboard.soundsheetmanagement.model;

import de.greenrobot.event.EventBus;
import org.neidhardt.dynamicsoundboard.DynamicSoundboardApplication;
import org.neidhardt.dynamicsoundboard.R;
import org.neidhardt.dynamicsoundboard.dao.DaoSession;
import org.neidhardt.dynamicsoundboard.dao.SoundSheet;
import org.neidhardt.dynamicsoundboard.misc.Util;
import org.neidhardt.dynamicsoundboard.navigationdrawer.soundlayouts.model.SoundLayoutsManager;
import org.neidhardt.dynamicsoundboard.navigationdrawer.soundsheets.events.SoundSheetsRemovedEvent;
import org.neidhardt.dynamicsoundboard.soundsheetmanagement.events.*;
import org.neidhardt.dynamicsoundboard.soundsheetmanagement.tasks.LoadSoundSheetsTask;
import org.neidhardt.dynamicsoundboard.soundsheetmanagement.tasks.StoreSoundSheetsTask;
import roboguice.util.SafeAsyncTask;

import java.util.ArrayList;
import java.util.List;

/**
 * File created by eric.neidhardt on 02.06.2015.
 */
public class SoundSheetsManager
		implements
			SoundSheetsDataAccess,
			SoundSheetsDataStorage,
			SoundSheetsDataUtil,
			OnOpenSoundSheetEventListener,
			OnSoundSheetsFromFileLoadedEventListener
{
	public static final String TAG = SoundSheetsManager.class.getName();

	private static final String DB_SOUND_SHEETS_DEFAULT = "org.neidhardt.dynamicsoundboard.soundsheet.SoundSheetManagerFragment.db_sound_sheets";
	private static final String DB_SOUND_SHEETS = "db_sound_sheets";

	List<SoundSheet> soundSheets;
	EventBus eventBus;

	private DaoSession daoSession;
	private boolean isInitDone;

	public SoundSheetsManager()
	{
		this.isInitDone = false;

		this.eventBus = EventBus.getDefault();
		this.init();
	}

	@Override
	public void registerOnEventBus()
	{
		if (!this.eventBus.isRegistered(this))
			this.eventBus.registerSticky(this, 1);
	}

	@Override
	public void unregisterOnEventBus()
	{
		this.eventBus.unregister(this);
	}

	@Override
	public void init()
	{
		if (this.isInitDone)
			throw new IllegalStateException(TAG + ": init() was called, but SoundSheetsManager was already initialized");

		this.isInitDone = true;

		this.soundSheets = new ArrayList<>();
		this.daoSession = Util.setupDatabase(DynamicSoundboardApplication.getSoundboardContext(), this.getDatabaseName());

		SafeAsyncTask task = new LoadSoundSheetsTask(this.daoSession, this);
		task.execute();
	}

	@Override
	public boolean isInit()
	{
		return this.isInitDone;
	}

	private String getDatabaseName()
	{
		String baseName = SoundLayoutsManager.getInstance().getActiveSoundLayout().getDatabaseId();
		if (baseName.equals(SoundLayoutsManager.DB_DEFAULT))
			return DB_SOUND_SHEETS_DEFAULT;
		return baseName + DB_SOUND_SHEETS;
	}

	@Override
	public List<SoundSheet> getSoundSheets()
	{
		if (this.soundSheets == null)
			this.soundSheets = new ArrayList<>();
		return this.soundSheets;
	}

	@Override
	public String addOrUpdateSoundSheet(SoundSheet soundSheet)
	{
		SoundSheet existingSoundSheet = null;
		int index = this.soundSheets.indexOf(soundSheet);
		if (index == -1)
			this.soundSheets.add(soundSheet);
		else
		{
			existingSoundSheet = this.soundSheets.get(index);
			existingSoundSheet.setFragmentTag(soundSheet.getFragmentTag());
			existingSoundSheet.setLabel(soundSheet.getLabel());
		}
		this.eventBus.post(new SoundSheetsChangedEvent());
		if (existingSoundSheet == null)
			this.eventBus.post(new OpenSoundSheetEvent(soundSheet));
		else
			this.eventBus.post(new OpenSoundSheetEvent(existingSoundSheet));

		if (existingSoundSheet != null)
			return existingSoundSheet.getFragmentTag();
		else
			return soundSheet.getFragmentTag();
	}

	@Override
	public SoundSheet getSoundSheetForFragmentTag(String fragmentTag)
	{
		for (SoundSheet soundSheet : this.soundSheets)
		{
			if (soundSheet.getFragmentTag().equals(fragmentTag))
				return soundSheet;
		}
		return null;
	}

	@Override
	public void setSelectedItem(int position)
	{
		int size = this.soundSheets.size();
		for (int i = 0; i < size; i++)
		{
			boolean isSelected = i == position;
			this.soundSheets.get(i).setIsSelected(isSelected);
		}
	}

	@Override
	public SoundSheet getSelectedItem()
	{
		for (SoundSheet soundSheet : this.soundSheets)
		{
			if (soundSheet.getIsSelected())
				return soundSheet;
		}
		return null;
	}

	@Override
	public void writeCacheBackAndRelease()
	{
		this.isInitDone = false;

		this.daoSession.getSoundSheetDao().deleteAll();
		SafeAsyncTask task = new StoreSoundSheetsTask(this.daoSession, this.soundSheets);
		task.execute();
	}

	@Override
	public String getSuggestedName()
	{
		int count = (this.soundSheets != null) ? this.soundSheets.size() : 0;
		return DynamicSoundboardApplication.getSoundboardContext().getResources().getString(R.string.suggested_sound_sheet_name) + count;
	}

	@Override
	public SoundSheet getNewSoundSheet(String label)
	{
		String tag = Integer.toString((label + DynamicSoundboardApplication.getRandomNumber()).hashCode());
		return new SoundSheet(null, tag, label, false);
	}

	@Override
	public void removeSoundSheet(SoundSheet soundSheet)
	{
		this.soundSheets.remove(soundSheet);
		this.eventBus.post(new SoundSheetsRemovedEvent(soundSheet));
	}

	@Override
	public void removeAllSoundSheets()
	{
		this.soundSheets.clear();
		this.eventBus.post(new SoundSheetsChangedEvent());
	}

	@Override
	public void onEvent(OpenSoundSheetEvent event)
	{
		int indexOfSelectedItem = this.soundSheets.indexOf(event.getSoundSheetToOpen());
		this.setSelectedItem(indexOfSelectedItem);
		EventBus.getDefault().post(new SoundSheetsChangedEvent());
	}

	@Override
	public void addLoadedSoundSheets(List<SoundSheet> soundSheetList)
	{
		this.soundSheets.addAll(soundSheetList);
		this.findSelectionAndDeselectOthers();
		this.eventBus.post(new SoundSheetsChangedEvent());
		this.eventBus.post(new SoundSheetsLoadedEvent(soundSheetList));
	}

	private void findSelectionAndDeselectOthers()
	{
		SoundSheet selected = null;
		if (this.soundSheets != null)
		{
			for (SoundSheet soundSheet : this.soundSheets)
			{
				if (soundSheet.getIsSelected() && selected == null)
					selected = soundSheet;
				else
					soundSheet.setIsSelected(false);
			}
		}
	}

	/**
	 * Called by LoadSoundSheetsTask when loading of SoundSheets has been finished.
	 * @param event delivered SoundSheetsLoadedEvent
	 */
	@SuppressWarnings("unused")
	public void onEventMainThread(SoundSheetsRemovedEvent event)
	{
		this.soundSheets.remove(event.getRemovedSoundSheet());
	}

	@Override
	public void onEvent(SoundSheetsFromFileLoadedEvent event)
	{
		// clear SoundSheets before adding new values
		// this removes all sounds in SoundSheets, but no in playlist
		this.soundSheets.clear();
		this.daoSession.getSoundSheetDao().deleteAll();

		this.soundSheets.addAll(event.getNewSoundSheetList());
		this.eventBus.post(new SoundSheetsChangedEvent());
	}

}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/soundsheetmanagement/model/SoundSheetsManager.kt
package org.neidhardt.dynamicsoundboard.soundsheetmanagement.model

import android.content.Context
import org.greenrobot.eventbus.EventBus
import org.neidhardt.dynamicsoundboard.R
import org.neidhardt.dynamicsoundboard.SoundboardApplication
import org.neidhardt.dynamicsoundboard.dao.DaoSession
import org.neidhardt.dynamicsoundboard.dao.SoundSheet
import org.neidhardt.dynamicsoundboard.dao.SoundSheetDao
import org.neidhardt.dynamicsoundboard.daohelper.GreenDaoHelper
import org.neidhardt.dynamicsoundboard.mediaplayer.PlaylistTAG
import org.neidhardt.dynamicsoundboard.misc.Logger
import org.neidhardt.dynamicsoundboard.soundlayoutmanagement.model.*
import org.neidhardt.dynamicsoundboard.soundsheetmanagement.events.OpenSoundSheetEvent
import org.neidhardt.dynamicsoundboard.soundsheetmanagement.events.SoundSheetAddedEvent
import org.neidhardt.dynamicsoundboard.soundsheetmanagement.events.SoundSheetChangedEvent
import org.neidhardt.dynamicsoundboard.soundsheetmanagement.events.SoundSheetsRemovedEvent
import org.neidhardt.dynamicsoundboard.soundsheetmanagement.tasks.loadSoundSheets
import rx.android.schedulers.AndroidSchedulers
import java.util.*

/**
 * File created by eric.neidhardt on 06.07.2015.
 */
class SoundSheetsManager(private val context: Context, private val soundLayoutsManager: ISoundLayoutManager) :
		SoundSheetsDataAccess,
		SoundSheetsDataStorage,
		SoundSheetsDataUtil
{
	private val TAG = javaClass.name

	private val DB_SOUND_SHEETS_DEFAULT = "org.neidhardt.dynamicsoundboard.soundsheet.SoundSheetManagerFragment.db_sound_sheets"
	private val DB_SOUND_SHEETS = "db_sound_sheets"

	private var daoSession: DaoSession? = null

	private val soundSheets: MutableList<SoundSheet> = ArrayList()
	private val eventBus = EventBus.getDefault()

	init {
		RxSoundLayoutManager.selectsLayout(this.soundLayoutsManager).subscribe { selectedLayout ->
			this.soundSheets.clear()
			this.daoSession = GreenDaoHelper.setupDatabase(this.context, this.getDatabaseName())
			this.daoSession?.let { dbSoundSheets ->
				SoundboardApplication.taskCounter.value += 1
				dbSoundSheets.loadSoundSheets()
						.flatMapIterable { it -> it }
						.observeOn(AndroidSchedulers.mainThread())
						.subscribe( { soundSheet ->
							Logger.d(TAG, "Loaded: $soundSheet")
							this.addSoundSheetToManager(soundSheet)
						}, { error ->
							Logger.e(TAG, "Error while loading sound sheets: ${error.message}")
							SoundboardApplication.taskCounter.value -= 1
						}, {
							Logger.d(TAG, "Loading sound sheets completed")
							SoundboardApplication.taskCounter.value -= 1
						} )
			}
		}
	}

	override fun releaseAll() {
		val copyList = ArrayList<SoundSheet>(soundSheets.size)
		copyList.addAll(soundSheets)

		this.soundSheets.clear()

		this.eventBus.post(SoundSheetsRemovedEvent(copyList))
	}

	private fun getDatabaseName(): String {
		val baseName = this.soundLayoutsManager.soundLayouts.activeLayout.databaseId
		if (baseName == SoundLayoutManager.DB_DEFAULT)
			return DB_SOUND_SHEETS_DEFAULT
		return baseName + DB_SOUND_SHEETS
	}

	override fun isPlaylistSoundSheet(fragmentTag: String): Boolean = fragmentTag == PlaylistTAG

	override fun getDbSoundSheets(): DaoSession = this.daoSession as DaoSession

	override fun getSoundSheets(): List<SoundSheet> = this.soundSheets

	override fun addSoundSheetToManager(soundSheet: SoundSheet) {
		this.soundSheets.add(soundSheet)

		val isSelected = soundSheet.isSelected

		if (isSelected)
			this.setSoundSheetSelected(soundSheet)

		soundSheet.insertItemInDatabaseAsync()

		this.eventBus.post(SoundSheetAddedEvent(soundSheet))
	}

	override fun removeSoundSheets(soundSheets: List<SoundSheet>) {
		val copyList = ArrayList<SoundSheet>(soundSheets.size)
		copyList.addAll(soundSheets) // this is done to prevent concurrent modification exception

		val dao = this.getDbSoundSheets().soundSheetDao
		for (soundSheetToRemove in copyList)
		{
			this.soundSheets.remove(soundSheetToRemove)
			if (soundSheetToRemove.isSelected)
			{
				if (this.soundSheets.size > 0)
					this.setSoundSheetSelected(this.soundSheets[0])

			}
			if (soundSheetToRemove.id != null)
				dao.delete(soundSheetToRemove)
			else
			{
				val list = dao.queryBuilder().where(SoundSheetDao.Properties.FragmentTag.eq(soundSheetToRemove.fragmentTag)).list()
				dao.deleteInTx(list)
			}
		}

		this.eventBus.post(SoundSheetsRemovedEvent(copyList))
	}

	override fun getSoundSheetForFragmentTag(fragmentTag: String): SoundSheet? {
		val results = this.soundSheets.filter { soundSheet -> soundSheet.fragmentTag == fragmentTag }
		return if (results.isNotEmpty()) results[0] else null
	}

	override fun setSoundSheetSelected(soundSheetToSelect: SoundSheet) {
		if (!this.soundSheets.contains(soundSheetToSelect))
			throw UnsupportedOperationException("$TAG: can not select SoundSheet $soundSheetToSelect because it is not loaded")

		for (soundSheet in this.soundSheets)
		{
			if (soundSheet != soundSheetToSelect && soundSheet.isSelected) // make sure only one item is selected
			{
				soundSheet.isSelected = false
				soundSheet.updateItemInDatabaseAsync()
				this.eventBus.post(SoundSheetChangedEvent(soundSheet))
			}
		}
		soundSheetToSelect.isSelected = true
		this.eventBus.post(SoundSheetChangedEvent(soundSheetToSelect))
		this.eventBus.post(OpenSoundSheetEvent(soundSheetToSelect))
	}

	override fun getSelectedItem(): SoundSheet? {
		val results = this.soundSheets.filter { soundSheet -> soundSheet.isSelected }
		return if (results.isNotEmpty()) results[0] else null
	}

	override fun getSuggestedName(): String = this.context.resources.getString(R.string.suggested_sound_sheet_name) + this.soundSheets.size

	override fun getNewSoundSheet(label: String): SoundSheet {
		val tag = Integer.toString((label + SoundboardApplication.randomNumber).hashCode())
		return SoundSheet(null, tag, label, false)
	}

}